newIDE/app/src/ProjectsStorage/CloudStorageProvider/CloudProjectWriter.js

```diff
@@ -0,0 +1,175 @@
+import { type FileMetadata } from '..';
+import {
+  commitVersion,
+  createCloudProject,
+  getCredentialsForProject,
+  updateCloudProject,
+} from '../../Utils/GDevelopServices/Project';
+import { serializeToJSON } from '../../Utils/Serializer';
+import { initializeZipJs } from '../../Utils/Zip.js';
+import CloudSaveAsDialog from './CloudSaveAsDialog';
+
+const zipProject = async (project: gdProject) => {
+  const zipJs: ZipJs = await initializeZipJs();
+  const projectJson = serializeToJSON(project);
+  const textReader = new zipJs.TextReader(projectJson);
+
+  return new Promise((resolve, reject) => {
+    zipJs.createWriter(new zipJs.BlobWriter('application/zip'), zipWriter => {
+      zipWriter.add('game.json', textReader, () => {
+        zipWriter.close(blob => {
+          resolve(blob);
+        });
+      });
+    });
+  });
+};
+
+const zipProjectAndCommitVersion = async ({
+  authenticatedUser,
+  project,
+  cloudProjectId,
+}: {
+  authenticatedUser: AuthenticatedUser,
+  project: gdProject,
+  cloudProjectId: string,
+}): Promise<?string> => {
+  const archive = await zipProject(project);
+  const newVersion = await commitVersion(
+    authenticatedUser,
+    cloudProjectId,
+    archive
+  );
+  return newVersion;
+};
+
+export const generateOnSaveProject = (
+  authenticatedUser: AuthenticatedUser
+) => async (project: gdProject, fileMetadata: FileMetadata) => {
+  const newVersion = await zipProjectAndCommitVersion({
+    authenticatedUser,
+    project,
+    cloudProjectId: fileMetadata.fileIdentifier,
+  });
+  if (!newVersion) return { wasSaved: false, fileMetadata };
+  return {
+    wasSaved: true,
+    fileMetadata,
+  };
+};
+
+export const generateOnChangeProjectProperty = (
+  authenticatedUser: AuthenticatedUser
+) => async (
+  project: gdProject,
+  fileMetadata: FileMetadata,
+  properties: { name: string }
+): Promise<boolean> => {
+  if (!authenticatedUser.authenticated) return false;
+  try {
+    await updateCloudProject(
+      authenticatedUser,
+      fileMetadata.fileIdentifier,
+      properties
+    );
+    const newVersion = await zipProjectAndCommitVersion({
+      authenticatedUser,
+      project,
+      cloudProjectId: fileMetadata.fileIdentifier,
+    });
+    if (!newVersion) {
+      throw new Error("Couldn't save project following property update.");
+    }
+
+    return true;
+  } catch (error) {
+    // TODO: Determine if a feedback should be given to user so that they can try again if necessary.
+    console.warn(
+      'An error occurred while changing cloud project name. Ignoring.',
+      error
+    );
+    return false;
+  }
+};
+
+const createCloudProjectAndInitCommit = async (
+  project: gdProject,
+  authenticatedUser: AuthenticatedUser,
+  name: string
+): Promise<?string> => {
+  try {
+    const cloudProject = await createCloudProject(authenticatedUser, {
+      name,
+    });
+    if (!cloudProject) return;
```

Comment:
Open question: should you throw instead of returning?

Reply:
For clarity's sake, it's better so I changed it. But in the end, it won't appear as an error to the user since it will return `wasSaved: false`